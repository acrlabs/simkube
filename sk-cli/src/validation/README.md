# SimKube Trace Validation Checks

| code | name | description |
|---|---|---|
| W0000 | status_field_populated |  Indicates that the status field of a Kubernetes object in the trace is non-empty; status fields are updated by their controlling objects and shouldn't be applied "by hand".  This is probably "fine" but it would be better to clean them up (and also they take up a lot of space.  |

This file is auto-generated; to rebuild, run `make validation_rules`.
