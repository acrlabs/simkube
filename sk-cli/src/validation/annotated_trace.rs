use std::collections::BTreeMap;

use sk_core::external_storage::{
    ObjectStoreWrapper,
    SkObjectStore,
};
use sk_store::{
    TraceEvent,
    TraceStorable,
    TraceStore,
};

#[derive(Clone, Default)]
pub struct AnnotatedTraceEvent {
    pub data: TraceEvent,
    pub annotations: Vec<(usize, String)>,
}

#[derive(Default)]
pub struct AnnotatedTrace {
    pub base: TraceStore,
    pub path: String,
    pub events: Vec<AnnotatedTraceEvent>,
    pub summary: BTreeMap<String, usize>,
}

impl AnnotatedTrace {
    pub async fn new(trace_path: &str) -> anyhow::Result<AnnotatedTrace> {
        let object_store = SkObjectStore::new(trace_path)?;
        let trace_data = object_store.get().await?.to_vec();
        let base = TraceStore::import(trace_data, &None)?;
        let events = base
            .iter()
            .map(|(event, _)| AnnotatedTraceEvent { data: event.clone(), ..Default::default() })
            .collect();
        Ok(AnnotatedTrace {
            base,
            events,
            path: trace_path.into(),
            ..Default::default()
        })
    }
}
