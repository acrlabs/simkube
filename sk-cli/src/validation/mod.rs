mod annotated_trace;
mod status_field_populated;
mod validation_store;
mod validator;

use clap::{
    value_parser,
    Subcommand,
    ValueEnum,
};
use sk_core::prelude::*;

pub use self::annotated_trace::AnnotatedTrace;
pub use self::validation_store::ValidationStore;

#[derive(Subcommand)]
pub enum ValidateSubcommand {
    #[command(about = "check a trace file")]
    Check(CheckArgs),

    #[command(about = "explain a rule")]
    Explain(ExplainArgs),

    #[command(about = "print all validation rules")]
    Print(PrintArgs),
}

#[derive(clap::Args)]
pub struct CheckArgs {
    #[arg(long_help = "location of the input trace file")]
    pub trace_path: String,
}

#[derive(clap::Args)]
pub struct ExplainArgs {
    #[arg(long_help = "Error code to explain")]
    pub code: String,
}

#[derive(Clone, ValueEnum)]
pub enum PrintFormat {
    Json,
    List,
    Table,
    Yaml,
}

#[derive(clap::Args)]
pub struct PrintArgs {
    #[arg(
        short,
        long,
        long_help = "format to display the validation rules",
        default_value = "list",
        value_parser = value_parser!(PrintFormat),
    )]
    pub format: PrintFormat,
}

pub async fn cmd(subcommand: &ValidateSubcommand) -> EmptyResult {
    let mut validators = ValidationStore::default();
    match subcommand {
        ValidateSubcommand::Check(args) => {
            let mut trace = AnnotatedTrace::new(&args.trace_path).await?;
            validators.validate_trace(&mut trace);
            print_summary(&trace, &validators)?;
        },
        ValidateSubcommand::Explain(args) => validators.explain(&args.code)?,
        ValidateSubcommand::Print(args) => validators.print(&args.format)?,
    }
    Ok(())
}

fn print_summary(trace: &AnnotatedTrace, validators: &ValidationStore) -> EmptyResult {
    for (code, count) in trace.summary_iter() {
        let name = validators.lookup(code)?.name;
        println!("{name} ({code}): {count:.>30}");
    }
    Ok(())
}

#[cfg(test)]
pub mod tests;
