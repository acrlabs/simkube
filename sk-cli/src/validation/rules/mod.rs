pub mod service_account_missing;
pub mod status_field_populated;

#[cfg(test)]
mod tests;
