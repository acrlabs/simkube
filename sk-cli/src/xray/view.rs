use std::iter::repeat;

use chrono::TimeDelta;
use ratatui::prelude::*;
use ratatui::widgets::{
    Block,
    Borders,
    Clear,
    List,
    Padding,
    Paragraph,
};
use sk_core::k8s::KubeResourceExt;
use sk_store::TraceStorable;

use super::app::{
    App,
    Mode,
};
use super::util::format_duration;

pub(super) fn view(app: &mut App, frame: &mut Frame) {
    let layout = Layout::default()
        .direction(Direction::Vertical)
        .constraints(vec![Constraint::Percentage(100), Constraint::Min(5)])
        .split(frame.area());
    let (top, bottom) = (layout[0], layout[1]);

    let events_border = Block::bordered().title(app.trace.path.clone());
    let object_border = Block::bordered();

    if top.width > 120 {
        let lr_layout = Layout::default()
            .direction(Direction::Horizontal)
            .constraints(vec![Constraint::Percentage(50), Constraint::Percentage(50)])
            .split(top);
        let (left, right) = (lr_layout[0], lr_layout[1]);

        render_event_list(app, frame, events_border.inner(left));
        if app.mode == Mode::EventSelected || app.mode == Mode::ObjectSelected {
            render_object(app, frame, object_border.inner(right));
        }

        frame.render_widget(events_border, left);
        frame.render_widget(object_border, right);
    } else {
        render_event_list(app, frame, events_border.inner(top));
        frame.render_widget(events_border, top);

        if app.mode == Mode::ObjectSelected {
            let popup = Rect {
                x: top.width / 10,
                y: top.height / 10,
                width: 4 * top.width / 5,
                height: 4 * top.height / 5,
            };
            frame.render_widget(Clear, popup);
            render_object(app, frame, object_border.inner(popup));
            frame.render_widget(object_border, popup);
        }
    }

    let greeting2 = Paragraph::new("Hello SimKube!\nUse arrows to navigate, space to select, 'q' to quit.")
        .white()
        .block(Block::new().borders(Borders::ALL));
    frame.render_widget(greeting2, bottom);
}

fn render_event_list(app: &mut App, frame: &mut Frame, layout: Rect) {
    // Here's some sortof obnoxious code; we'd like to have the event list "expand" so that you can
    // see the applied and deleted objects for that particular event.  The way we do this is split
    // our layout into three sublayouts; the first includes all the events up to the selected one,
    // then we nest in one level and display the applied and deleted objects, then we unnest and
    // display the rest of the events
    let num_events = app.trace.events.len();
    let start_ts = app.trace.base.start_ts().unwrap_or(0);

    // Add one so the selected event is included on top
    let (sel_index_inclusive, sel_event) = match app.mode {
        Mode::EventSelected | Mode::ObjectSelected => {
            let sel_index = app.event_list_state.selected().unwrap();
            (sel_index + 1, Some(&app.trace.events[sel_index]))
        },
        _ => (num_events, None),
    };

    let mut root_items_1 = Vec::with_capacity(sel_index_inclusive);
    let mut root_items_2 = Vec::with_capacity(num_events - sel_index_inclusive);

    for (i, evt) in app.trace.events.iter().enumerate() {
        let d = TimeDelta::new(evt.data.ts - start_ts, 0).unwrap();
        let d_str = format!(
            "{} ({} applied/{} deleted)",
            format_duration(d),
            evt.data.applied_objs.len(),
            evt.data.deleted_objs.len()
        );
        if i < sel_index_inclusive {
            root_items_1.push(d_str);
        } else {
            root_items_2.push(d_str);
        }
    }

    let sublist_items = sel_event.map_or(vec![], |evt| {
        let mut items: Vec<_> = evt
            .data
            .applied_objs
            .iter()
            .zip(repeat("+"))
            .chain(evt.data.deleted_objs.iter().zip(repeat("-")))
            .map(|(obj, op)| format!("  {} {}", op, obj.namespaced_name()))
            .collect();
        if items.is_empty() {
            items.push(String::new());
        }
        items
    });

    let nested_layout = Layout::default()
        .direction(Direction::Vertical)
        .constraints(vec![
            // We know how many lines we have; use max constraints here so the lists are next to
            // each other.  The last one can be min(0) and take up the rest of the space
            Constraint::Max(root_items_1.len() as u16),
            Constraint::Max(sublist_items.len() as u16),
            Constraint::Min(0),
        ])
        .split(layout);

    let list_part_one = List::new(root_items_1)
        .highlight_style(Style::new().add_modifier(Modifier::REVERSED))
        .highlight_symbol(">> ");
    let sublist = List::new(sublist_items)
        .highlight_style(Style::new().bg(Color::Blue))
        .highlight_symbol("++ ")
        .style(Style::new().italic());
    let list_part_two = List::new(root_items_2).block(Block::new().padding(Padding::left(3)));

    frame.render_stateful_widget(list_part_one, nested_layout[0], &mut app.event_list_state);
    frame.render_stateful_widget(sublist, nested_layout[1], &mut app.object_list_state);
    frame.render_widget(list_part_two, nested_layout[2])
}

fn render_object(app: &mut App, frame: &mut Frame, layout: Rect) {
    let evt_idx = app.event_list_state.selected().unwrap();
    let obj_idx = app.object_list_state.selected().unwrap();
    let applied_len = app.trace.events[evt_idx].data.applied_objs.len();
    let deleted_len = app.trace.events[evt_idx].data.deleted_objs.len();

    let obj = if obj_idx >= applied_len {
        if obj_idx - applied_len > deleted_len {
            return;
        }
        &app.trace.events[evt_idx].data.deleted_objs[obj_idx - applied_len]
    } else {
        &app.trace.events[evt_idx].data.applied_objs[obj_idx]
    };

    let obj_str = serde_json::to_string_pretty(obj).unwrap();
    let contents = List::new(obj_str.split('\n')).highlight_style(Style::new().bg(Color::Blue));
    frame.render_stateful_widget(contents, layout, &mut app.object_contents_list_state);
}
