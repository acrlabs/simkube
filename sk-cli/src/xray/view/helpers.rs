use chrono::TimeDelta;
use lazy_static::lazy_static;
use ratatui::prelude::*;
use sk_core::prelude::*;

use crate::validation::{
    AnnotatedTraceEvent,
    ValidatorType,
};

pub(super) const LIST_PADDING: usize = 3;
lazy_static! {
    static ref ERR_STYLE: Style = Style::new().white().on_red();
    static ref WARN_STYLE: Style = Style::new().white().on_yellow();
}

pub(super) fn make_event_spans(event: &AnnotatedTraceEvent, start_ts: i64) -> (Span, Span) {
    let d = TimeDelta::new(event.data.ts - start_ts, 0).unwrap();
    let evt_span = Span::from(format!(
        "{} ({} applied/{} deleted)",
        format_duration(d),
        event.data.applied_objs.len(),
        event.data.deleted_objs.len()
    ));

    let (warnings, errs) = event.annotations.iter().fold((0, 0), |(mut w, mut e), codes| {
        for code in codes {
            match code.0 {
                ValidatorType::Warning => w += 1,
                ValidatorType::Error => e += 1,
            }
        }
        (w, e)
    });

    if warnings + errs == 0 {
        (evt_span, Span::default())
    } else {
        let err_str = format!("{} error{}", errs, if errs == 1 { "" } else { "s" });
        let warn_str = format!("{} warning{}", warnings, if warnings == 1 { "" } else { "s" });
        let style = if errs == 0 { *WARN_STYLE } else { *ERR_STYLE };
        let err_span = Span::styled(format!(" {err_str}/{warn_str} "), style);

        (evt_span, err_span)
    }
}

pub(super) fn make_object_spans<'a>(
    index: usize,
    obj: &DynamicObject,
    op: char,
    event: &'a AnnotatedTraceEvent,
) -> (Span<'a>, Span<'a>) {
    let obj_span = Span::styled(format!("  {} {}", op, obj.namespaced_name()), Style::new().italic());
    let (warnings, errs) = event.annotations[index].iter().fold((0, 0), |(mut w, mut e), code| {
        match code.0 {
            ValidatorType::Warning => w += 1,
            ValidatorType::Error => e += 1,
        }
        (w, e)
    });
    if warnings + errs == 0 {
        (obj_span, Span::default())
    } else {
        let err_str = format!("{} error{}", errs, if errs == 1 { "" } else { "s" });
        let warn_str = format!("{} warning{}", warnings, if warnings == 1 { "" } else { "s" });
        let style = if errs == 0 { *WARN_STYLE } else { *ERR_STYLE };
        let err_span = Span::styled(format!(" {err_str}/{warn_str} "), style);

        (obj_span, err_span)
    }
}

pub(super) fn format_list_entries<'a>(
    spans: impl Iterator<Item = (Span<'a>, Span<'a>)> + Clone,
    width: usize,
) -> Vec<Text<'a>> {
    let max_err_width = spans
        .clone()
        .max_by_key(|(_, err_span)| err_span.width())
        .map_or(0, |(_, err_span)| err_span.width());

    spans
        .map(|(mut item_span, err_span)| {
            // Instead of using a table we're just computing everything by hand here.  The format
            // of a list entry is LIST_PADDING|item|spaces|error|LIST_PADDING
            //
            // In order to ensure that all the error lines are the same width, we compute the
            // maximum above, and then use that everywhere.
            //
            // If the length of the item + the error + 2 * LIST_PADDING > width of the screen,
            // we have no spaces in the middle.  That's what these saturating subs are computing.
            // On the other hand, if there's no error to display, then we just fill out the rest of
            // the line with spaces.
            let mid_padding_width = width
                .saturating_sub(item_span.width())
                .saturating_sub(max_err_width)
                .saturating_sub(LIST_PADDING * 2);
            let mid_padding_span = Span::from(" ".repeat(mid_padding_width));

            // We want the right padding to match the same style as the error or warning, so if an
            // error is present we create a separate span here.
            let (trunc_width, right_padding_width) = match err_span.width() {
                0 => (width - LIST_PADDING, 0),
                x => (
                    width - (mid_padding_width + max_err_width + max_err_width - x + LIST_PADDING + LIST_PADDING),
                    max_err_width - x + LIST_PADDING,
                ),
            };
            let right_padding_span = Span::styled(" ".repeat(right_padding_width), err_span.style);

            // If the list entry is too long, then we truncate it and replace the rightmost
            // characters with an ellipsis.
            if trunc_width < item_span.content.len() {
                let content = item_span.content.to_mut();
                content.truncate(trunc_width);
                content.replace_range(trunc_width - 4..trunc_width, "... ");
            }

            Text::from(item_span.clone() + mid_padding_span + err_span.clone() + right_padding_span)
        })
        .collect()
}

pub(super) fn format_duration(d: TimeDelta) -> String {
    let day_str = match d.num_days() {
        x if x > 0 => format!("{x}d "),
        _ => String::new(),
    };

    format!("{}{:02}:{:02}:{:02}", day_str, d.num_hours() % 24, d.num_minutes() % 60, d.num_seconds() % 60)
}
