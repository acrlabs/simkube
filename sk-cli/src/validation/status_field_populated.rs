use super::annotated_trace::AnnotatedTraceEvent;
use super::validation_store::{
    Diagnostic,
    Validator,
    ValidatorType,
};

const HELP: &str = r#"Indicates that the status field of a Kubernetes object in
the trace is non-empty; status fields are updated by their controlling objects
and shouldn't be applied "by hand".  This is probably "fine" but it would be
better to clean them up (and also they take up a lot of space."#;

#[derive(Default)]
pub(super) struct StatusFieldPopulated {}

impl Diagnostic for StatusFieldPopulated {
    fn check_next_event(&mut self, event: &mut AnnotatedTraceEvent) -> Vec<usize> {
        event
            .data
            .applied_objs
            .iter()
            .enumerate()
            .filter_map(|(i, obj)| {
                if obj
                    .data
                    .as_object()
                    .expect("DynamicObject data should be a map")
                    .get("status")
                    .is_some_and(|v| !v.is_null())
                {
                    return Some(i);
                }
                None
            })
            .collect()
    }

    fn reset(&mut self) {}
}

pub(super) fn validator() -> Validator {
    Validator {
        type_: ValidatorType::Warning,
        name: "status_field_populated",
        help: HELP,
        diagnostic: Box::new(StatusFieldPopulated::default()),
    }
}
