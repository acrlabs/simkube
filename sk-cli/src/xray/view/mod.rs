mod helpers;

use std::cmp::{
    max,
    min,
};
use std::iter::{
    once,
    repeat,
};

use ratatui::prelude::*;
use ratatui::widgets::{
    Block,
    Borders,
    Clear,
    List,
    ListState,
    Padding,
    Paragraph,
};

use self::helpers::*;
use super::app::{
    App,
    JumpDir,
    Mode,
};

const MIN_TWO_PANEL_WIDTH: u16 = 120;

pub(super) fn view(app: &mut App, frame: &mut Frame) {
    let layout = Layout::default()
        .direction(Direction::Vertical)
        .constraints(vec![Constraint::Percentage(100), Constraint::Min(5)])
        .split(frame.area());
    let (top, bottom) = (layout[0], layout[1]);

    let events_border = Block::bordered().title(app.annotated_trace.path());
    let object_border = Block::bordered();
    let (events_inner, object_inner);

    if top.width > MIN_TWO_PANEL_WIDTH {
        let lr_layout = Layout::default()
            .direction(Direction::Horizontal)
            .constraints(vec![Constraint::Percentage(50), Constraint::Percentage(50)])
            .split(top);
        let (left, right) = (lr_layout[0], lr_layout[1]);

        events_inner = events_border.inner(left);
        object_inner = object_border.inner(right);
        render_event_list(app, frame, events_inner);
        if app.mode == Mode::EventSelected || app.mode == Mode::ObjectSelected {
            render_object(app, frame, object_inner);
        }

        frame.render_widget(events_border, left);
        frame.render_widget(object_border, right);
    } else {
        events_inner = events_border.inner(top);
        render_event_list(app, frame, events_inner);
        frame.render_widget(events_border, top);

        // We compute the popup dimensions so we can cleanly forward-declare object_inner
        // above, but we don't actually render anything unless we're in the right mode.
        let popup = Rect {
            x: top.width / 10,
            y: top.height / 10,
            width: 4 * top.width / 5,
            height: 4 * top.height / 5,
        };
        object_inner = object_border.inner(popup);

        if app.mode == Mode::ObjectSelected {
            frame.render_widget(Clear, popup);
            render_object(app, frame, object_inner);
            frame.render_widget(object_border, popup);
        }
    }

    let greeting2 = Paragraph::new("Hello SimKube!\nUse arrows to navigate, space to select, 'q' to quit.")
        .white()
        .block(Block::new().borders(Borders::ALL));
    frame.render_widget(greeting2, bottom);
}

fn render_event_list(app: &mut App, frame: &mut Frame, layout: Rect) {
    // Here's some sortof obnoxious code; we'd like to have the event list "expand" so that you can
    // see the applied and deleted objects for that particular event.  The way we do this is split
    // our layout into three sublayouts; the first includes all the events up to the selected one,
    // then we nest in one level and display the applied and deleted objects, then we unnest and
    // display the rest of the events
    let num_events = app.annotated_trace.len();
    let start_ts = app.annotated_trace.start_ts().unwrap_or(0);

    // Add one so the selected event is included on top
    let hi_index = app.highlighted_event_index();
    let (sel_index_inclusive, sel_event) = match app.mode {
        Mode::EventSelected | Mode::ObjectSelected => (hi_index + 1, app.annotated_trace.get_event(hi_index)),
        _ => (num_events, None),
    };

    let event_spans = app.annotated_trace.iter().map(|event| make_event_spans(event, start_ts));
    let mut top_entries = format_list_entries(event_spans, layout.width as usize);
    let bottom_entries = top_entries.split_off(sel_index_inclusive);

    // chain together the applied and deleted objects with either a "+" or "-" prefix, respectively
    let obj_spans = sel_event.map_or(vec![], |evt| {
        let mut sublist_items = evt
            .data
            .applied_objs
            .iter()
            .zip(repeat('+'))
            .chain(evt.data.deleted_objs.iter().zip(repeat('-')))
            .enumerate()
            .map(|(i, (obj, op))| make_object_spans(i, obj, op, evt))
            .peekable();
        if sublist_items.peek().is_none() {
            // if there are no objects associated with this event, we display an empty span
            format_list_entries(once((Span::default(), Span::default())), layout.width as usize)
        } else {
            format_list_entries(sublist_items, layout.width as usize)
        }
    });

    // Compute the constraint values for the first part of the events list and the objects list
    // AND ALSO update the selected and offset pointers if we've jumped; we have to interweave
    // these because we base the jump distance on the currently viewed events and objects
    let (top_height, mid_height) = if app.mode == Mode::RootView {
        if let Some(dir) = app.jump {
            jump_list_state(&mut app.event_list_state, dir, top_entries.len(), layout.height, false);
        }
        // If we're in the root view, the number of entries to display is just the total number of
        // events minus the current view offset.  Since there is no selected event, the number of
        // middle entries is 0.
        (top_entries.len().saturating_sub(app.event_list_state.offset()) as u16, 0)
    } else {
        // If we've selected an event to view, we have to first compute the height ot the top list
        // as above; the length of the middle list is either:
        //   - 1 if the selected event is at the very bottom of the display (which will push the offset of
        //     the top list up by one)
        //   - the number of rows between the selected event and the bottom of the display, OR
        //   - the total number of objects belonging to this event (if they all fit)
        let th = top_entries.len().saturating_sub(app.event_list_state.offset()) as u16;
        let mh = min(
            max(1, layout.height.saturating_sub(th)),
            obj_spans.len().saturating_sub(app.object_list_state.offset()) as u16,
        );

        // Once we know how many objects we can display for this event, we can compute the amount
        // to jump (if any); the key observation that makes this all work is that if we've selected
        // an event, the length of the top entries is fixed.
        if app.mode == Mode::EventSelected {
            if let Some(dir) = app.jump {
                jump_list_state(&mut app.object_list_state, dir, obj_spans.len(), mh, true);
            };
        }
        (th, mh)
    };

    // We know how many lines we have; use max constraints here so the lists are next to
    // each other.  The last one can be min(0) and take up the rest of the space
    let nested_layout = Layout::default()
        .direction(Direction::Vertical)
        .constraints(vec![Constraint::Max(top_height), Constraint::Max(mid_height), Constraint::Min(0)])
        .split(layout);

    let list_part_one = List::new(top_entries)
        .highlight_style(Style::new().add_modifier(Modifier::REVERSED))
        .highlight_symbol(">> ");
    let sublist = List::new(obj_spans)
        .highlight_style(Style::new().bg(Color::Blue))
        .highlight_symbol("++ ");
    let list_part_two = List::new(bottom_entries).block(Block::new().padding(Padding::left(LIST_PADDING as u16)));

    frame.render_stateful_widget(list_part_one, nested_layout[0], &mut app.event_list_state);
    frame.render_stateful_widget(sublist, nested_layout[1], &mut app.object_list_state);
    frame.render_widget(list_part_two, nested_layout[2]);
}

fn render_object(app: &mut App, frame: &mut Frame, layout: Rect) {
    let event_idx = app.event_list_state.selected().unwrap();
    let obj_idx = app.object_list_state.selected().unwrap();

    let Some(obj) = app.annotated_trace.get_object(event_idx, obj_idx) else {
        return;
    };
    let obj_str = serde_json::to_string_pretty(obj).unwrap();
    let contents = List::new(obj_str.split('\n')).highlight_style(Style::new().bg(Color::Blue));

    if app.mode == Mode::ObjectSelected {
        if let Some(dir) = app.jump {
            jump_list_state(&mut app.object_contents_list_state, dir, contents.len(), layout.height, false);
        }
    }

    frame.render_stateful_widget(contents, layout, &mut app.object_contents_list_state);
}

fn jump_list_state(list_state: &mut ListState, jump_dir: JumpDir, list_len: usize, view_height: u16, pin_bottom: bool) {
    // compute how far to jump in the specified direction, given the number of items in the
    // list that we're trying to display and how much room we have to display them
    let offset = list_state.offset();
    let selected = list_state.selected().unwrap();

    let new_pos = match jump_dir {
        // If we jump down, we increase the selection by the view height - 1, so that the
        // last-visible item before the jump becomes the first-visible item after the jump
        JumpDir::Down => (offset + view_height as usize).saturating_sub(1),

        // If we jump up, and we aren't on the first item, just jump to the top of the current
        // page; otherwise, page up, but subtract one so that the top item before the jump
        // becomes the bottom item after the jump
        JumpDir::Up => {
            if selected != offset {
                offset
            } else {
                // make sure the -1 is _inside_ the saturating_sub, instead of adding 1 _outside_;
                // otherwise you get weird behaviours when selected == offset == 0.
                offset.saturating_sub(view_height as usize - 1)
            }
        },
    };

    // we're relying on the ratatui behaviour of adjusting the selected index to be "within bounds"
    // if we computed something out of bounds above.
    list_state.select(Some(new_pos));

    // pin_bottom means that we don't want to have empty space at the bottom if we jumped "too
    // far", so we make a special (smaller) offset computation in this case
    if pin_bottom && list_len.saturating_sub(new_pos) < view_height as usize {
        *list_state.offset_mut() = list_len.saturating_sub(view_height as usize);
    } else if new_pos < list_len {
        // otherwise, we just set the new offset to the selected item, so that it appears at the
        // top of the page
        *list_state.offset_mut() = new_pos;
    }
}

#[cfg(test)]
mod tests;
