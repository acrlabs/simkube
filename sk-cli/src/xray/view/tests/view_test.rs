use std::fs::File;
use std::io::BufReader;

use insta::assert_debug_snapshot;
use ratatui::backend::TestBackend;
use ratatui::prelude::*;
use ratatui::widgets::ListState;
use sk_store::ExportedTrace;

use super::*;
use crate::validation::tests::{
    annotated_trace,
    test_validation_store,
};
use crate::validation::{
    AnnotatedTrace,
    AnnotatedTraceEvent,
    ValidationStore,
};
use crate::xray::view::jump_list_state;

#[fixture]
fn test_app(test_validation_store: ValidationStore, mut annotated_trace: AnnotatedTrace) -> App {
    test_validation_store.validate_trace(&mut annotated_trace, false).unwrap();
    App {
        annotated_trace,
        event_list_state: ListState::default().with_selected(Some(0)),
        ..Default::default()
    }
}

#[fixture]
fn test_app_large(test_validation_store: ValidationStore) -> App {
    let trace_data_file = File::open("../testdata/large_trace.json").unwrap();
    let reader = BufReader::new(trace_data_file);
    let exported_trace: ExportedTrace = serde_json::from_reader(reader).unwrap();
    let annotated_trace = AnnotatedTrace::new_with_events(
        exported_trace
            .events()
            .iter()
            .cloned()
            .map(|e| AnnotatedTraceEvent::new(e))
            .collect(),
    );

    test_app(test_validation_store, annotated_trace)
}

#[rstest]
#[case::top(0, 0, false, 19, 19)]
#[case::top_mid_sel(10, 0, false, 19, 19)]
#[case::top_pin(0, 0, true, 19, 19)]
#[case::mid(20, 20, false, 39, 39)]
// the view height is 20, we are trying to jump to 39 but this would leave 17 empty spaces
// since there are only 42 elements in the list, so instead we set the offset to 22
#[case::mid_pin(20, 20, true, 39, 22)]
// when rendered, the "selected" index will become the last index in the list but for now we
// just naively add the viewport height to it, which is why the expected offset is 60 here
#[case::bottom(41, 41, false, 60, 41)]
#[case::bottom_pin(41, 41, true, 60, 22)]
fn test_jump_list_state_down(
    #[case] selected: usize,
    #[case] offset: usize,
    #[case] pin: bool,
    #[case] expected_selected: usize,
    #[case] expected_offset: usize,
) {
    let mut list_state = ListState::default().with_offset(offset).with_selected(Some(selected));
    jump_list_state(&mut list_state, JumpDir::Down, 42, 20, pin);
    assert_eq!(list_state.offset(), expected_offset);
    assert_eq!(list_state.selected().unwrap(), expected_selected);
}

#[rstest]
#[case::top(0, 0, 0, 0)]
#[case::top_mid_sel(10, 0, 0, 0)]
#[case::mid_to_top(10, 10, 0, 0)]
#[case::mid(23, 23, 4, 4)]
#[case::mid_mid_sel(25, 23, 23, 23)]
#[case::bottom(42, 42, 23, 23)]
fn test_jump_list_state_up(
    #[case] selected: usize,
    #[case] offset: usize,
    #[case] expected_offset: usize,
    #[case] expected_selected: usize,
) {
    let mut list_state = ListState::default().with_offset(offset).with_selected(Some(selected));
    jump_list_state(&mut list_state, JumpDir::Up, 42, 20, false);
    assert_eq!(list_state.offset(), expected_offset);
    assert_eq!(list_state.selected().unwrap(), expected_selected);
}

#[rstest]
#[case::first(0)]
#[case::last(3)]
fn itest_render_event_list(mut test_app: App, #[case] index: usize) {
    set_snapshot_suffix!("{index}");
    test_app.event_list_state.select(Some(index));
    let mut term = Terminal::new(TestBackend::new(80, 20)).unwrap();
    let cf = term.draw(|frame| view(&mut test_app, frame)).unwrap();
    assert_debug_snapshot!(cf);
}

#[rstest]
#[case::first(0)]
#[case::middle(2)]
#[case::last(3)]
fn itest_render_event_list_event_selected(mut test_app: App, #[case] index: usize) {
    set_snapshot_suffix!("{index}");
    test_app.mode = Mode::EventSelected;
    test_app.event_list_state.select(Some(index));
    test_app.object_list_state.select(Some(0));
    let mut term = Terminal::new(TestBackend::new(80, 20)).unwrap();
    let cf = term.draw(|frame| view(&mut test_app, frame)).unwrap();
    assert_debug_snapshot!(cf);
}

#[rstest]
#[case::top_1(0, 0)]
#[case::top_2(10, 0)]
#[case::bottom_1(80, 80)]
#[case::bottom_2(88, 80)]
#[case::bottom_3(88, 88)]
fn itest_render_large_event_list(mut test_app_large: App, #[case] selected: usize, #[case] offset: usize) {
    set_snapshot_suffix!("{selected}.{offset}");
    test_app_large.event_list_state.select(Some(selected));
    *test_app_large.event_list_state.offset_mut() = offset;
    let mut term = Terminal::new(TestBackend::new(80, 20)).unwrap();
    let cf = term.draw(|frame| view(&mut test_app_large, frame)).unwrap();
    assert_debug_snapshot!(cf);
}

#[rstest]
#[case::top_1(1, 0)]
#[case::top_2(3, 0)]
#[case::bottom_1(80, 80)]
#[case::bottom_2(88, 80)]
#[case::bottom_3(88, 88)]
fn itest_render_large_event_list_event_selected(
    mut test_app_large: App,
    #[case] selected: usize,
    #[case] offset: usize,
) {
    set_snapshot_suffix!("{selected}.{offset}");
    test_app_large.mode = Mode::EventSelected;
    test_app_large.event_list_state.select(Some(selected));
    *test_app_large.event_list_state.offset_mut() = offset;
    let mut term = Terminal::new(TestBackend::new(80, 20)).unwrap();
    let cf = term.draw(|frame| view(&mut test_app_large, frame)).unwrap();
    assert_debug_snapshot!(cf);
}
