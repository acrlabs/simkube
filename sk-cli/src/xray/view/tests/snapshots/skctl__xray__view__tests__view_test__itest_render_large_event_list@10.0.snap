---
source: sk-cli/src/xray/view/tests/view_test.rs
expression: cf
---
CompletedFrame {
    buffer: Buffer {
        area: Rect { x: 0, y: 0, width: 80, height: 20 },
        content: [
            "┌──────────────────────────────────────────────────────────────────────────────┐",
            "│   00:00:00 (0 applied/0 deleted)                                             │",
            "│   00:00:00 (27 applied/0 deleted)                      1 error/0 warnings    │",
            "│   00:07:03 (1 applied/0 deleted)                                             │",
            "│   00:09:08 (3 applied/0 deleted)                       1 error/0 warnings    │",
            "│   00:09:11 (1 applied/0 deleted)                                             │",
            "│   00:09:13 (1 applied/0 deleted)                                             │",
            "│   00:09:16 (109 applied/0 deleted)                     1 error/0 warnings    │",
            "│   00:09:23 (1 applied/0 deleted)                                             │",
            "│   00:09:43 (1 applied/0 deleted)                                             │",
            "│   00:12:48 (1 applied/0 deleted)                                             │",
            "│>> 00:13:09 (1 applied/0 deleted)                                             │",
            "│   00:13:17 (1 applied/0 deleted)                                             │",
            "│   00:13:32 (1 applied/0 deleted)                                             │",
            "└──────────────────────────────────────────────────────────────────────────────┘",
            "┌──────────────────────────────────────────────────────────────────────────────┐",
            "│Hello SimKube!                                                                │",
            "│Use arrows to navigate, space to select, 'q' to quit.                         │",
            "│                                                                              │",
            "└──────────────────────────────────────────────────────────────────────────────┘",
        ],
        styles: [
            x: 0, y: 0, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
            x: 56, y: 2, fg: White, bg: Red, underline: Reset, modifier: NONE,
            x: 79, y: 2, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
            x: 56, y: 4, fg: White, bg: Red, underline: Reset, modifier: NONE,
            x: 79, y: 4, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
            x: 56, y: 7, fg: White, bg: Red, underline: Reset, modifier: NONE,
            x: 79, y: 7, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
            x: 1, y: 11, fg: Reset, bg: Reset, underline: Reset, modifier: REVERSED,
            x: 79, y: 11, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
            x: 0, y: 15, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        ]
    },
    area: Rect {
        x: 0,
        y: 0,
        width: 80,
        height: 20,
    },
    count: 0,
}
